File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/01_introduction_to_algorithms/ES6/02_recursive_binary_search.js
/**
 * Searches recursively number from the list
 * @param {Array} list
 * @param {number} item Search item
 * @param {number} low Lower limit of search in the list
 * @param {number} high Highest limit of search in the list
 * @return {(number | null)} Number if the value is found or NULL otherwise
 */
const binarySearch = ( list, item, low = 0, high = list.length - 1 ) => {
    let mid = Math.floor((low + high) / 2);
    let guess = list[mid];

    if ( low > high ) return null;

    if ( guess === item ) {
        return mid;
    } else if ( guess > item ) {
        high = mid - 1;
        return binarySearch( list, item, low, high );
    } else {
        low = mid + 1;
        return binarySearch( list, item, low, high );
    }
};

/**
 * Creates the array that contains numbers 1...N
 * @param {number} n - number N
 * @return {Array}
 */
const createArr = ( n ) => Array.from({length: n}, (v, k) => k + 1);

const myList = createArr( 100 );

console.log( binarySearch( myList, 3 ) ); // 2
console.log( binarySearch( myList, -1 ) ); // null

File: notes-n-resources/Data-Structures-N-Algo/_DS-n-Algos/__MY_OPRIGINAL_DS/_Extra-Practice/01_introduction_to_algorithms/ruby/02_recursive_binary_search.rb
def binary_search(list, item, low = 0, high = list.length - 1)
  while low <= high
    mid = (low + high) / 2
    guess = list[mid]

    if guess == item
      return mid
    elsif guess > item
      high = mid - 1
      binary_search(list, item, low, high)
    else
      low = mid + 1
      binary_search(list, item, low, high)
    end
  end

  # If item is not found
  return nil
end

# Create array with numbers 1 through 100
my_list = (1..100).to_a

puts binary_search(my_list, 2) # => 1
puts binary_search(my_list, 50) # => 49
p binary_search(my_list, 101) # => nil
